### ISSUE TITLE
Different images processing methods in train_batch and val_batch.

### ISSUE THREAD

**Issue by zyiiiiiii** [ultralytics/ultralytics#9973]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone, I am using yolov8_obb to train my own custom data set, and in the process of observing training and validation, I find that the processing of images in the training and validation process seems to be very different. Is it because of the use of data enhancement, the images in the training look like this? However, will the difference in image processing between the training process and the validation process affect the performance of the model? Because there are some rotating targets that have not been well detected, that is, the box and the target do not fit well. Can you can give me some guidance, thanks in advance.

### Additional

images in training process:
![train_batch0](https://github.com/ultralytics/ultralytics/assets/166489453/93599397-591f-4b41-901a-89010740c1d6)
images in validation process:
![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/166489453/51d51713-ee88-4962-8a95-b279a461310b)

**Comment by glenn-jocher** [comment]
Hi there! 🌟

Indeed, the images used during the training and validation phases go through different processing. During training, data augmentation techniques such as rotation, scaling, and color space adjustments are heavily applied. This diversification helps the model learn to generalize better across a wide range of inputs.

For validation, however, the aim is to assess the model's performance on data as close to the original distribution as possible. Hence, minimal to no augmentation is applied, keeping the images more or less 'as is'. This discrepancy you're observing is intentional and is a common practice in machine learning to balance between model robustness and evaluation reliability.

Regarding the issue with rotating targets not being well detected, it could be related to the specific augmentations applied or the model's architecture suitability for your data. YOLOv8-OBB is designed to handle oriented bounding boxes, but performance can vary based on dataset characteristics.

One suggestion is to carefully review and possibly fine-tune the data augmentation strategies used during training. You might want to experiment with reducing some of the augmentations or introducing new ones more aligned with your data's specific challenges. Also, ensure your custom dataset is labeled accurately, especially the rotated objects, to provide the model with the best chance to learn from them.

For adjusting augmentation strategies, you can modify the data configuration file (usually in YAML format) associated with your training process. Here's a small code snippet on how to adjust an augmentation setting in a YAML file:

```yaml
# Example augmentation setting adjustment in data.yaml
augmentations:
  rotate: 45  # Rotates images between -45 and 45 degrees
  scale: 1.1   # Scales images up to 10%
```

Adjust the parameters according to your needs. Remember, finding the right balance in data augmentation is key to improving model performance on complex datasets such as ones with rotated objects.

Hope this helps, and happy training!

**Comment by zyiiiiiii** [comment]
@glenn-jocher Thank you for your detail answer, regard to train rotated objects, what parameters do I need to pay more attention to, because there are  a lot of hyper parameters in yolov8_obb model. Could you give me some guidance？

**Comment by zyiiiiiii** [comment]
@glenn-jocher 
1.About scale, when I set scale=1.1, there is a error, that is scale should be between 0 and 1.
2.Where is anchor ratios parameter and how to set a custom anchors to representative my objects? Because my dataset objects have different types and sizes, some are rectangle, some are square, and for the square objects their sizes are different, some are bigger, some are smaller.
3.About optimizer, Optimizer=auto(default), what are the options other than auto? When I try to set to other optimizers, such as SGD, Adam and so on, will have a error.

**Comment by glenn-jocher** [comment]
Hi there! 😊

It looks like we've got a few things to iron out, so let's get right to it:

1. **Scale Issue:** The `scale` parameter indeed operates within the range [0, 1] to control the randomness of scaling applied to your images. If you're experiencing issues, try adjusting it within this range. Here's a quick fix example:
   ```yaml
   scale: 0.2  # Randomly scales images up to 20%
   ```

2. **Anchor Ratios and Custom Anchors:** Anchors are crucial for detecting objects of various shapes and sizes. You'll find your current anchors defined in your model's YAML file. For custom anchors tailored to your dataset, you can recalculate them based on your dataset-specific objects. Use the `--save-anchors` option during training, and then update your model's YAML file with the new anchors. Here's an example line from a YAML file where you might adjust anchors:
   ```yaml
   anchors: 3  # Number of anchor groups
   ```

3. **Optimizers:** The `optimizer` option defaults to 'auto', which chooses the best optimizer based on your setup. Other options include `'SGD'`, `'Adam'`, or `'AdamW'`. If you're encountering errors when switching optimizers, ensure you're specifying them as strings in your config YAML or command line. For instance:
   ```yaml
   optimizer: 'SGD'
   ```
   
Hope this helps clear things up! Keep experimenting, and you'll find the best settings for your unique dataset. 🚀

**Comment by zyiiiiiii** [comment]
@glenn-jocher Thanks for your detail reply and sorry to bother you again,  I didn't find anchors defined in model's YAML file, the following is the default setup of the YAML file:

# Ultralytics YOLO 🚀, AGPL-3.0 license
# Default training settings and hyperparameters for medium-augmentation COCO training

task: detect # (str) YOLO task, i.e. detect, segment, classify, pose
mode: train # (str) YOLO mode, i.e. train, val, predict, export, track, benchmark

# Train settings -------------------------------------------------------------------------------------------------------
model: # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data: # (str, optional) path to data file, i.e. coco128.yaml
epochs: 100 # (int) number of epochs to train for
time: # (float, optional) number of hours to train for, overrides epochs if supplied
patience: 100 # (int) epochs to wait for no observable improvement for early stopping of training
batch: 16 # (int) number of images per batch (-1 for AutoBatch)
imgsz: 640 # (int | list) input images size as int for train and val modes, or list[w,h] for predict and export modes
save: True # (bool) save train checkpoints and predict results
save_period: -1 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: False # (bool) True/ram, disk or False. Use cache for data loading
device: # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8 # (int) number of worker threads for data loading (per RANK if DDP)
project: # (str, optional) project name
name: # (str, optional) experiment name, results saved to 'project/name' directory
exist_ok: False # (bool) whether to overwrite existing experiment
pretrained: True # (bool | str) whether to use a pretrained model (bool) or a model to load weights from (str)
optimizer: auto # (str) optimizer to use, choices=[SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, auto]
verbose: True # (bool) whether to print verbose output
seed: 0 # (int) random seed for reproducibility
deterministic: True # (bool) whether to enable deterministic mode
single_cls: False # (bool) train multi-class data as single-class
rect: False # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: False # (bool) use cosine learning rate scheduler
close_mosaic: 10 # (int) disable mosaic augmentation for final epochs (0 to disable)
resume: False # (bool) resume training from last checkpoint
amp: True # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0 # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False # (bool) profile ONNX and TensorRT speeds during training for loggers
freeze: None # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
multi_scale: False # (bool) Whether to use multi-scale during training
# Segmentation
overlap_mask: True # (bool) masks should overlap during training (segment train only)
mask_ratio: 4 # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0 # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True # (bool) validate/test during training
split: val # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_json: False # (bool) save results to JSON file
save_hybrid: False # (bool) save hybrid version of labels (labels + additional predictions)
conf: # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.7 # (float) intersection over union (IoU) threshold for NMS
max_det: 300 # (int) maximum number of detections per image
half: False # (bool) use half precision (FP16)
dnn: False # (bool) use OpenCV DNN for ONNX inference
plots: True # (bool) save plots and images during train/val

# Predict settings -----------------------------------------------------------------------------------------------------
source: # (str, optional) source directory for images or videos
vid_stride: 1 # (int) video frame-rate stride
stream_buffer: False # (bool) buffer all streaming frames (True) or return the most recent frame (False)
visualize: False # (bool) visualize model features
augment: False # (bool) apply image augmentation to prediction sources
agnostic_nms: False # (bool) class-agnostic NMS
classes: # (int | list[int], optional) filter results by class, i.e. classes=0, or classes=[0,2,3]
retina_masks: False # (bool) use high-resolution segmentation masks
embed: # (list[int], optional) return feature vectors/embeddings from given layers

# Visualize settings ---------------------------------------------------------------------------------------------------
show: False # (bool) show predicted images and videos if environment allows
save_frames: False # (bool) save predicted individual video frames
save_txt: False # (bool) save results as .txt file
save_conf: False # (bool) save results with confidence scores
save_crop: False # (bool) save cropped images with results
show_labels: True # (bool) show prediction labels, i.e. 'person'
show_conf: True # (bool) show prediction confidence, i.e. '0.99'
show_boxes: True # (bool) show prediction boxes
line_width: # (int, optional) line width of the bounding boxes. Scaled to image size if None.

# Export settings ------------------------------------------------------------------------------------------------------
format: torchscript # (str) format to export to, choices at https://docs.ultralytics.com/modes/export/#export-formats
keras: False # (bool) use Kera=s
optimize: False # (bool) TorchScript: optimize for mobile
int8: False # (bool) CoreML/TF INT8 quantization
dynamic: False # (bool) ONNX/TF/TensorRT: dynamic axes
simplify: False # (bool) ONNX: simplify model
opset: # (int, optional) ONNX: opset version
workspace: 4 # (int) TensorRT: workspace size (GB)
nms: False # (bool) CoreML: add NMS

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.01 # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01 # (float) final learning rate (lr0 * lrf)
momentum: 0.937 # (float) SGD momentum/Adam beta1
weight_decay: 0.0005 # (float) optimizer weight decay 5e-4
warmup_epochs: 3.0 # (float) warmup epochs (fractions ok)
warmup_momentum: 0.8 # (float) warmup initial momentum
warmup_bias_lr: 0.1 # (float) warmup initial bias lr
box: 7.5 # (float) box loss gain
cls: 0.5 # (float) cls loss gain (scale with pixels)
dfl: 1.5 # (float) dfl loss gain
pose: 12.0 # (float) pose loss gain
kobj: 1.0 # (float) keypoint obj loss gain
label_smoothing: 0.0 # (float) label smoothing (fraction)
nbs: 64 # (int) nominal batch size
hsv_h: 0.015 # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.7 # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.4 # (float) image HSV-Value augmentation (fraction)
degrees: 0.0 # (float) image rotation (+/- deg)
translate: 0.1 # (float) image translation (+/- fraction)
scale: 0.5 # (float) image scale (+/- gain)
shear: 0.0 # (float) image shear (+/- deg)
perspective: 0.0 # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.0 # (float) image flip up-down (probability)
fliplr: 0.5 # (float) image flip left-right (probability)
mosaic: 1.0 # (float) image mosaic (probability)
mixup: 0.0 # (float) image mixup (probability)
copy_paste: 0.0 # (float) segment copy-paste (probability)
auto_augment: randaugment # (str) auto augmentation policy for classification (randaugment, autoaugment, augmix)
erasing: 0.4 # (float) probability of random erasing during classification training (0-1)
crop_fraction: 1.0 # (float) image crop fraction for classification evaluation/inference (0-1)

# Custom config.yaml ---------------------------------------------------------------------------------------------------
cfg: # (str, optional) for overriding defaults.yaml

# Tracker settings ------------------------------------------------------------------------------------------------------
tracker: botsort.yaml # (str) tracker type, choices=[botsort.yaml, bytetrack.yaml]

**Comment by glenn-jocher** [comment]
Hey there! No bother at all, I'm here to help 😊. The `anchors` aren't specified directly in the default YAML configuration file for the YOLOv8 models, including `yolov8_obb`. They're dynamically adjusted during training based on the dataset and model architecture.

However, if you're looking to customize or utilize specific anchors for your dataset (especially considering your varied object shapes and sizes), you can calculate them based on your dataset's bounding box distributions. Here's a way to do it:

1. **Calculate Custom Anchors:**
   You'd typically run a script or use a tool that analyzes your dataset's bounding boxes to suggest optimal anchors. While Ultralytics YOLOv8 automatically adjusts anchors during training, if you have a specific requirement or want to experiment, you might consider using external tools or scripts specifically designed for anchor calculation.

2. **Update Your Model's YAML:**
   Once you have your calculated anchors, you can manually specify them in your model's YAML configuration under the `anchors` field, although this is more common in earlier versions like YOLOv3 or YOLOv4.

3. **Optimizers:**
   The `optimizer: auto` setting dynamically chooses the best optimizer for your training scenario based on your model configuration. While YOLOv8 supports various optimizers like SGD, Adam, etc., if you're facing issues specifying them, ensure you're using the correct syntax in your training command or script. Typically, Ultralytics automatically manages this, but for experimentation, direct assignment in the script or command may be needed.

For the diverse sizes and types of objects in your dataset, focusing on a proper data augmentation strategy, and ensuring your training dataset is representative of your task's complexity, will significantly impact performance.

Lastly, don't hesitate to explore Ultralytics resources, like discussions and issue trackers, as they contain a wealth of information from the community. Happy training! 🚀

**Comment by zyiiiiiii** [comment]
@glenn-jocher Hi, I still have some questions:

1.In the train_batch0, train_batch1 and train_batch2, why does some part of the images shift, such as the last row of the image in the following figure（train_batch1）, and I found that all the images in the last row are like this?  I have checked the relevant parameters of the data enhancement and found that "translate" seems to be related to the translation of the images, but I have set the value of this parameter to 0, why do some images still shift?  What parameters cause this to happen in the images? Thanks in advance.

train_batch1:
![train_batch1](https://github.com/ultralytics/ultralytics/assets/166489453/ad457f3b-42e7-4de1-9a0f-f3f026bbf962)

**Comment by zyiiiiiii** [comment]
@glenn-jocher Hi, I still have some questions:

Question 1:
When the objects are square, is it challenge for YOLOv8_OBB to detect them? Because I found that when the rotational objects are square most of the predict outputs are not rotational, but for the rectangle objects, the predict results are look well.

Question 2:
If Question 1 is true, how to overcome this challenge? Thanks in advance.
